takanon_extractions/dynamic_extractions/process_document.py

```diff
@@ -0,0 +1,466 @@
+"""
+
+import argparse
+import sys
+import subprocess
+from datetime import datetime
+from pathlib import Path
+from typing import Optional, List, Dict, Any
+import json
+import time
+
+from botnim.config import get_logger
+from pipeline_config import PipelineConfig, PipelineMetadata, PipelineStage, Environment, validate_json_structure
+
+# Logger setup
+logger = get_logger(__name__)
+
+class PipelineRunner:
+    """Main pipeline orchestrator."""
+    
+    def __init__(self, config: PipelineConfig):
+        self.config = config
+        self.metadata = PipelineMetadata()
+        self.start_time = None
+        
+    def run(self) -> bool:
+        """Run the complete pipeline."""
+        logger.info("Starting HTML processing pipeline")
+        logger.info(f"Configuration: {self.config.to_dict()}")
+        
+        # Validate configuration
+        errors = self.config.validate()
+        if errors:
+            logger.error("Configuration validation failed:")
+            for error in errors:
+                logger.error(f"  - {error}")
+            return False
+        
+        # Initialize metadata
+        self.start_time = datetime.now()
+        self.metadata.start_time = self.start_time.isoformat()
+        
+        try:
+            # Stage 1: Extract structure
+            if not self._run_structure_extraction():
+                return False
+            
+            # Stage 2: Extract content
+            if not self._run_content_extraction():
+                return False
+            
+            # Stage 3: Generate markdown files
+            if not self._run_markdown_generation():
+                return False
+            
+            # Pipeline completed successfully
+            self.metadata.end_time = datetime.now().isoformat()
+            self._save_pipeline_metadata()
+            
+            logger.info("Pipeline completed successfully")
+            self._print_summary()
+            return True
+            
+        except Exception as e:
+            logger.error(f"Pipeline failed with exception: {e}")
+            self.metadata.errors.append(f"Pipeline exception: {str(e)}")
+            self.metadata.end_time = datetime.now().isoformat()
+            self._save_pipeline_metadata()
+            return False
+    
+    def _run_structure_extraction(self) -> bool:
+        """Run structure extraction stage."""
+        logger.info("Stage 1: Extracting document structure")
+        
+        stage_start = time.time()
+        
+        # Check if output already exists
+        if self.config.structure_file.exists() and not self.config.overwrite_existing:
+            logger.warning(f"Structure file already exists: {self.config.structure_file}")
+            if not self._confirm_overwrite():
+                logger.info("Skipping structure extraction")
+                self.metadata.stages_completed.append(PipelineStage.EXTRACT_STRUCTURE)
+                return True
+        
+        # Build command
```

Comment:
Please import the function from the file and run it instead of using subprocess, because:
1) Simplify code
2) Easier to pass arguments (e.g. can pass the config instance directly)
3) Allows code analysis to detect mismatches between caller and called

Reply:
@noamoss ☝️

Reply:
I hope I understood it correctly. If so, fixed - for all stages.